### PR TITLE
add new separateKustomizeComponent and skipMembers fields

### PR DESCRIPTION
**Pull request by MatousJobanek**
add two new fields into kubesaw-admins.yaml file
1. `separateKustomizeComponent` can be used in a member cluster config - when set to true, then the manifests for the member cluster is gonna be generated in a separate Kustomize component/folder
2. `selector.skipMembers` which can be defined either for a ServiceAccount or for a Users. It's a list of members the generation should be skipped for the given SA or User entity.

These two fields together enable us to generate manifests for a given member cluster (the intel hackathon one) in a separate folder and skip some users/sa there. With this, we can still generate manifests needed to execute ksctl commands in the intel hackathon cluster, but it won't include the manifests for the crtadmin users of the 3rd party components (the components won't be installed there).

NOTE: You may ask why did I use the "selector" as a parent field for the "skipMembers" one. I'd like to move the `AllClusters` field there as well and manage any other fields related to filtering/selecting the items under this generic one..
(But if it would be too confusing, then I can remove the parent field and store it all at the root level of the entity)

To be done in the next PR:
* adapt this change in also in the `generate cli-configs` command

### CODE REVIEW

File: pkg/cmd/generate/mock_test.go [hunk 1]
```diff
@@ -67,9 +67,21 @@ func newAdminManifestsContext(t *testing.T, config *assets.KubeSawAdmins, files
...
 	}
+	for _, modify := range options {
+		modify(ctx)
+	}
+	return ctx
+}
+
+type fakeClusterContextOption func(ctx *clusterContext)
+
+func withSpecialKMemberName(specialKMemberName string) fakeClusterContextOption {
```
Comment (mfrancisc) [inline pkg/cmd/generate/mock_test.go:83]: very minor - but the field name seems to be `specific` not `special`, should we use that terminology ?

**Review by mfrancisc — COMMENTED** [review]
Looks good! 

I have left only few minor comments.

File: pkg/cmd/generate/admin-manifests.go [hunk 2]
```diff
@@ -87,12 +103,17 @@ type adminManifestsContext struct {
 	files         assets.FS
 }
 
-func ensureCluster(ctx *adminManifestsContext, clusterType configuration.ClusterType, cache objectsCache) error {
-	ctx.PrintContextSeparatorf("Generating manifests for %s cluster type", clusterType)
+func ensureCluster(ctx *adminManifestsContext, clusterType configuration.ClusterType, cache objectsCache, specificKMemberName string) error {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cmd/generate/admin-manifests.go:106]: That was intentional, it's a shortcut for specificKustomizeMemberName or specificKustomizeComponentMemberName

File: pkg/cmd/generate/mock_test.go (hunk 1 above)
Comment (MatousJobanek) [reply pkg/cmd/generate/mock_test.go]: good catch, looks like there is a mix of the formats - unified in f677a45

File: pkg/cmd/generate/admin-manifests_test.go [hunk 3]
```diff
@@ -65,19 +67,43 @@ func TestAdminManifests(t *testing.T) {
...
+			require.NoError(t, err)
+			term := NewFakeTerminalWithResponse("Y")
+			term.Tee(os.Stdout)
+			flags := newAdminManifestsFlags(outDir(outTempDir), kubeSawAdminsFile(configFile), singleCluster())
+
+			// when
+			err = adminManifests(term, files, flags)
+
+			// then
+			require.Error(t, err)
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cmd/generate/admin-manifests_test.go]: good idea - done in bd9e8d7

File: pkg/assets/sandbox_config.go [hunk 4]
```diff
@@ -12,25 +12,32 @@ type Clusters struct {
...
 type ServiceAccount struct {
-	Name                      string `yaml:"name"`
-	Namespace                 string `yaml:"namespace,omitempty"`
+	Name                      string   `yaml:"name"`
+	Namespace                 string   `yaml:"namespace,omitempty"`
+	Selector                  Selector `yaml:"selector"`
 	PermissionsPerClusterType `yaml:",inline"`
 }
 
+type Selector struct {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/assets/sandbox_config.go:33]: I'm bad at comments :man_facepalming: 
added comments here 4a2c567

File: pkg/cmd/generate/admin-manifests.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mfrancisc) [reply pkg/cmd/generate/admin-manifests.go:106]: oh gotcha! thanks !

**Review by mfrancisc — APPROVED** [review]
Great job 👍 

Thanks for addressing my comments.

**Review by alexeykazakov — APPROVED** [review]

**Review by xcoulon — COMMENTED** [review]
I'm a bit confused with the new `MemberCluster.SeparateKustomizeComponent` boolean field.
The comment on the field says:

> when set to true, then the manifests for the member cluster will be generated in a separate Kustomize folder

but how/where do we specify which Kustomize folder? 
and what do you mean by "Kustomize folder", by the way?

**Comment by MatousJobanek** [comment]
that's exactly why I created the other PR so you can see what it looks like: https://github.com/codeready-toolchain/sandbox-sre/pull/1926

>  and what do you mean by "Kustomize folder", by the way?

a folder that contains the whole structure of the generated manifests with the kustomization.yaml files


Let me improve the comment there

**Comment by xcoulon** [comment]
I'm a bit confused with the new 

> that's exactly why I created the other PR so you can see what it looks like: [codeready-toolchain/sandbox-sre#1926](https://github.com/codeready-toolchain/sandbox-sre/pull/1926)
> 

yes, I've seen it, but I'm still a bit confused :/ 
Not by the fact that the generated manifests will be in a separate folder (here, `components/auth/devsandbox-production/generated-manifests/member-intel`), but rather, how to we specify the path to this particular folder?

> > and what do you mean by "Kustomize folder", by the way?
> 
> a folder that contains the whole structure of the generated manifests with the kustomization.yaml files
> 
> Let me improve the comment there

may I suggest "Kustomize output folder"?

**Comment by MatousJobanek** [comment]
> Not by the fact that the generated manifests will be in a separate folder (here, components/auth/devsandbox-production/generated-manifests/member-intel), but rather, how to we specify the path to this particular folder?

I'm confused, what do you want to have different there?

**Comment by xcoulon** [comment]
> I updated the comment [7719c2a](https://github.com/kubesaw/ksctl/commit/7719c2a2450c1daf0f3bd711d57ba63959c8e937) (before seeing your last message) please take a look and feel free to make a suggestion in the GH PR view

great, thanks for taking my questions/comments into account!

**Review by xcoulon — APPROVED** [review]

**Pull request merged**